README.md
# twemproxy (nutcracker) [![Build Status](https://secure.travis-ci.org/twitter/twemproxy.png)](http://travis-ci.org/twitter/twemproxy)

**twemproxy** (pronounced "two-em-proxy"), aka **nutcracker** is a fast and lightweight proxy for [memcached](http://www.memcached.org/) and [redis](http://redis.io/) protocol. It was primarily built to reduce the connection count on the backend caching servers.

## Build

To build nutcracker from [distribution tarball](http://code.google.com/p/twemproxy/downloads/list):

    $ ./configure
    $ make
    $ sudo make install

To build nutcracker from [distribution tarball](http://code.google.com/p/twemproxy/downloads/list) in _debug mode_:

    $ CFLAGS="-ggdb3 -O0" ./configure --enable-debug=full
    $ make
    $ sudo make install

To build nutcracker from source with _debug logs enabled_ and _assertions enabled_:

    $ git clone git@github.com:twitter/twemproxy.git
    $ cd twemproxy
    $ autoreconf -fvi
    $ ./configure --enable-debug=full
    $ make
    $ src/nutcracker -h

Few checklists:

+ use newer version of gcc (older version of gcc has problems)
+ use CFLAGS="-O1" ./configure && make
+ use CFLAGS="-O3 -fno-strict-aliasing" ./configure && make

## New Features

+ New command: replace_server(now just for redis).
+ Log rotate: we add three startup items:-R(--log-rotate), -M(--log-file-max-size) and -C(--log-file-count). If you want use this function, use -R to run the nutcracker. -M requires a file size between 1000000 and 1122601371959296 bytes(1PB). -C requires an integer between -1 and 200, -1 means do not delete old log files, 0 means delete all old log files, N(0<N<=200) means retain N old log files and old log files naming:logfilename_time.
+ Replication pool: we allow one server pool as another server pool's replication pool(now just for memcache and limit two pools Recursive). we add three keys for the configuration file:replication_from, replication_mode and write_back_mode. It allows double write for master pool and slave pool. If get/gets miss in master pool,  penetrate to the slave pool, and if result hit in the slave pool, write back to master pool.  
+ Tcp keepalive: add tcp keepalive for twemproxy.
+ Administration: add administration for twemproxy. Use telnet to connect administer.
+ Configuration reload: let twemproxy can reload configuration on runtime.
+ Zookeeper: add zookeeper for twemproxy. twemproxy can start and keep with zookeeper.

## Features

+ Fast.
+ Lightweight.
+ Maintains persistent server connections.
+ Keeps connection count on the backend caching servers low.
+ Enables pipelining of requests and responses.
+ Supports proxying to multiple servers.
+ Supports multiple server pools simultaneously.
+ Shard data automatically across multiple servers.
+ Implements the complete [memcached ascii](notes/memcache.txt) and [redis](notes/redis.md) protocol.
+ Easy configuration of server pools through a YAML file.
+ Supports multiple hashing modes including consistent hashing and distribution.
+ Can be configured to disable nodes on failures.
+ Observability through stats exposed on stats monitoring port.
+ Works with Linux, *BSD, OS X and Solaris (SmartOS)

## Help

	Usage: nutcracker [-?hVdDt] [-v verbosity level] [-o output file]
					  [-c conf file] [-s stats port] [-a stats addr]
					  [-i stats interval] [-p pid file] [-m mbuf size]

	Options:
	  -h, --help                : this help
	  -V, --version             : show version and exit
	  -t, --test-conf           : test configuration for syntax errors and exit
	  -d, --daemonize           : run as a daemon
	  -D, --describe-stats      : print stats description and exit
	  -v, --verbosity=N         : set logging level (default: 5, min: 0, max: 11)
	  -o, --output=S            : set logging file (default: stderr)
	  -c, --conf-file=S         : set configuration file (default: conf/nutcracker.yml)
	  -s, --stats-port=N        : set stats monitoring port (default: 22222)
	  -a, --stats-addr=S        : set stats monitoring ip (default: 0.0.0.0)
	  -i, --stats-interval=N    : set stats aggregation interval in msec (default: 30000 msec)
	  -p, --pid-file=S          : set pid file (default: off)
	  -m, --mbuf-size=N         : set size of mbuf chunk in bytes (default: 16384 bytes)
	  
	  -R, --log-rorate          : enable log rorate
	  -M, --log-file-max-size=N : set max size per log file if log-rorate open (default: 1073741824 bytes, min: 1000000 bytes, max: 1122601371959296 bytes)
	  -C, --log-file-count=N    : set total count of log files to leave if log-rorate open (default: 10, min: -1, max: 200)
	  -A, --proxy-adm-addr=S    : set proxy administer monitoring ip (default: 0.0.0.0)
	  -P, --proxy-adm-port=N    : set proxy administer monitoring port (default: 0, means disabled)
	  -S, --zk-start            : set configuration from zookeeper
	  -K, --zk-keep             : set configuration keep with zookeeper
	  -z, --zk-server=S         : set zookeeper servers address (default: 127.0.0.1:2181)
	  -Z, --zk-path=S           : set zookeeper configuration path (default: /twemproxy)
  
## Zero Copy

In nutcracker, all the memory for incoming requests and outgoing responses is allocated in mbuf. Mbuf enables zero-copy because the same buffer on which a request was received from the client is used for forwarding it to the server. Similarly the same mbuf on which a response was received from the server is used for forwarding it to the client.

Furthermore, memory for mbufs is managed using a reuse pool. This means that once mbuf is allocated, it is not deallocated, but just put back into the reuse pool. By default each mbuf chunk is set to 16K bytes in size. There is a trade-off between the mbuf size and number of concurrent connections nutcracker can support. A large mbuf size reduces the number of read syscalls made by nutcracker when reading requests or responses. However, with large mbuf size, every active connection would use up 16K bytes of buffer which might be an issue when nutcracker is handling large number of concurrent connections from clients. When nutcracker is meant to handle a large number of concurrent client connections, you should set chunk size to a small value like 512 bytes using the -m or --mbuf-size=N argument.

## Configuration

nutcracker can be configured through a YAML file specified by the -c or --conf-file command-line argument on process start. The configuration file is used to specify the server pools and the servers within each pool that nutcracker manages. The configuration files parses and understands the following keys:

+ **listen**: The listening address and port (name:port or ip:port) for this server pool.
+ **hash**: The name of the hash function. Possible values are:
 + one_at_a_time
 + md5
 + crc16
 + crc32 (crc32 implementation compatible with [libmemcached](http://libmemcached.org/))
 + crc32a (correct crc32 implementation as per the spec)
 + fnv1_64
 + fnv1a_64
 + fnv1_32
 + fnv1a_32
 + hsieh
 + murmur
 + jenkins
+ **hash_tag**: A two character string that specifies the part of the key used for hashing. Eg "{}" or "$$". [Hash tag](notes/recommendation.md#hash-tags)  enable mapping different keys to the same server as long as the part of the key within the tag is the same.
+ **distribution**: The key distribution mode. Possible values are:
 + ketama
 + modula
 + random
+ **timeout**: The timeout value in msec that we wait for to establish a connection to the server or receive a response from a server. By default, we wait indefinitely.
+ **backlog**: The TCP backlog argument. Defaults to 512.
+ **preconnect**: A boolean value that controls if nutcracker should preconnect to all the servers in this pool on process start. Defaults to false.
+ **redis**: A boolean value that controls if a server pool speaks redis or memcached protocol. Defaults to false.
+ **server_connections**: The maximum number of connections that can be opened to each server. By default, we open at most 1 server connection.

ChangeLog
2015-07-07  deep
	
	*twemproxy: version 0.4.8 release
	 replace_server(redis)
	 log rotate
	 tcp keepalive
	 administration
	 configuration reload
	 zookeeper
	 replication pool(mc)

2014-18-10  idning <idning@gmail.com>

    * twemproxy: version 0.4.0 release
      mget improve (idning)
      many new commands supported: LEX, PFADD, PFMERGE, SORT, PING, QUIT, SCAN... (mattrobenolt, areina, idning)
      handle max open file limit(allenlz)
      add notice-log and use ms time in log(idning)
      fix bug in string_compare (andyqzb)
      fix deadlock in sighandler (idning)

2013-20-12  Manju Rajashekhar  <manj@cs.stanford.edu>
    * twemproxy: version 0.3.0 release
      SRANDMEMBER support for the optional count argument (mkhq)
      Handle case where server responds while the request is still being sent (jdi-tagged)
      event ports (solaris/smartos) support
      add timestamp when the server was ejected
      support for set ex/px/nx/xx for redis 2.6.12 and up (ypocat)
      kqueue (bsd) support (ferenyx)
      fix parsing redis response to accept integer reply (charsyam)

2013-23-04  Manju Rajashekhar  <manj@cs.stanford.edu>
    * twemproxy: version 0.2.4 release
      redis keys must be less than mbuf_data_size() in length (fifsky)
      Adds support for DUMP/RESTORE commands in Redis (remotezygote)
      Use of the weight value in the modula distribution (mezzatto)
      Add support to unix socket connections to servers (mezzatto)
      only check for duplicate server name and not 'host:port:weight' when 'name' is configured
      crc16 hash support added (mezzatto)

2013-31-01  Manju Rajashekhar  <manj@twitter.com>
    * twemproxy: version 0.2.3 release
      RPOPLPUSH, SDIFF, SDIFFSTORE, SINTER, SINTERSTORE, SMOVE, SUNION, SUNIONSTORE, ZINTERSTORE, and ZUNIONSTORE support (dcartoon)
      EVAL and EVALSHA support (ferenyx)
      exit 1 if configuration file is invalid (cofyc)
      return non-zero exit status when nutcracker cannot start for some reason
      use server names in stats (charsyam)
      Fix failure to resolve long FQDN name resolve (conmame)
      add support for hash tags

2012-18-10  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.2.2 release
      fix the off-by-one error when calculating redis key length

2012-12-10  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.2.1 release
      don't use buf in conf_add_server
      allow an optional instance name for consistent hashing (charsyam)
      add --stats-addr=S option
      add stats-bind-any -a option (charsyam)

2012-12-03  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.2.0 release
      add -D or --describe-stats command-line argument to print stats description
      redis support in twemproxy
      setup pre/post splitcopy and pre/post coalesce handlers in msg struct
      memcache pre_splitcopy, post_splitcopy, pre_coalesce and post_coalesce handlers
      every fragment of a msg vector keeps track of the first/last fragment, number of fragments and fragment owner
      set up msg parser handler for memcache connections
      refactor parsing code and create header file nc_proto.h
      stats_listen should use st->addr as the listening address string
      delete stats tracking memcache requests and responses; stats module no longer tracks protocol related stats

2012-10-27  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.1.20 release
      on msg_repair, msg->pos should point to nbuf->pos and not nbuf->last
      refactor memcache parsing code into proto directory
      add redis option to configuration file
      fix macro definition strXcmp error for big endian
      fix log_hexdump and loga_hexdump

2012-07-31  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.1.19 release
      close server connection on a stray response (yashh, bmatheny)

2012-06-19  Manju Rajashekhar  <manj@twitter.com>

    * twemproxy: version 0.1.18 release
      command line option to set mbuf chunk size


src/nc.c
/*
 * twemproxy - A fast and lightweight proxy for memcached protocol.
 * Copyright (C) 2011 Twitter, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <getopt.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/utsname.h>

#include <nc_core.h>
#include <nc_conf.h>
#include <nc_signal.h>

#define NC_CONF_PATH        "conf/nutcracker.yml"

#define NC_LOG_DEFAULT      LOG_NOTICE
#define NC_LOG_MIN          LOG_EMERG
#define NC_LOG_MAX          LOG_PVERB
#define NC_LOG_PATH         NULL

#define NC_STATS_PORT       STATS_PORT
#define NC_STATS_ADDR       STATS_ADDR
#define NC_STATS_INTERVAL   STATS_INTERVAL

#define NC_PID_FILE         NULL

#define NC_MBUF_SIZE        MBUF_SIZE
#define NC_MBUF_MIN_SIZE    MBUF_MIN_SIZE
#define NC_MBUF_MAX_SIZE    MBUF_MAX_SIZE

#if 1 //shenzheng 2015-1-26 log rotating
#define NC_LOG_ROTATE_DEFAULT			LOG_ROTATE_DEFAULT

#define NC_LOG_FILE_MAX_SIZE_DEFAULT	LOG_FILE_MAX_SIZE_DEFAULT
#define NC_LOG_FILE_MAX_SIZE_MIN		LOG_FILE_MAX_SIZE_MIN
#define NC_LOG_FILE_MAX_SIZE_MAX		LOG_FILE_MAX_SIZE_MAX

#define NC_LOG_FILE_COUNT_DEFAULT		LOG_FILE_COUNT_DEFAULT
#define NC_LOG_FILE_COUNT_MIN			LOG_FILE_COUNT_MIN
#define NC_LOG_FILE_COUNT_MAX			LOG_FILE_COUNT_MAX
#endif //shenzheng 2015-1-26 log rotating

#if 1 //shenzheng 2015-4-28 proxy administer
#define NC_PROXY_ADM_PORT       PROXY_ADM_PORT
#define NC_PROXY_ADM_ADDR       PROXY_ADM_ADDR
#endif //shenzheng 2015-4-28 proxy administer

#if 1 //shenzheng 2015-6-8 zookeeper
#ifdef NC_ZOOKEEPER
#define NC_ZOOKEEPER_START_DEFAULT      ZOOKEEPER_START_DEFAULT
#define NC_ZOOKEEPER_KEEP_DEFAULT       ZOOKEEPER_KEEP_DEFAULT
#define NC_ZOOKEEPER_ADDR       		ZOOKEEPER_ADDR
#define NC_ZOOKEEPER_PATH       		ZOOKEEPER_PATH
#endif
#endif //shenzheng 2015-6-8 zookeeper

static int show_help;
static int show_version;
static int test_conf;
static int daemonize;
static int describe_stats;

static struct option long_options[] = {
    { "help",           	no_argument,        NULL,   'h' },
    { "version",        	no_argument,        NULL,   'V' },
    { "test-conf",      	no_argument,        NULL,   't' },
    { "daemonize",      	no_argument,        NULL,   'd' },
    { "describe-stats", 	no_argument,        NULL,   'D' },
    { "verbose",        	required_argument,  NULL,   'v' },
    { "output",         	required_argument,  NULL,   'o' },
    { "conf-file",      	required_argument,  NULL,   'c' },
    { "stats-port",     	required_argument,  NULL,   's' },
    { "stats-interval", 	required_argument,  NULL,   'i' },
    { "stats-addr",     	required_argument,  NULL,   'a' },
    { "pid-file",       	required_argument,  NULL,   'p' },
    { "mbuf-size",      	required_argument,  NULL,   'm' },
#if 1 //shenzheng 2015-1-26 log rotating
	{ "log-rorate",     	no_argument,  		NULL,   'R' },
	{ "log-file-max-size",  required_argument,  NULL,   'M' },
	{ "log-file-count",     required_argument,  NULL,   'C' },
#endif //shenzheng 2015-1-26 log rotating
#if 1 //shenzheng 2015-4-28 proxy administer
	{ "proxy-adm-addr",     required_argument,  NULL,   'A' },
	{ "proxy-adm-port",     required_argument,  NULL,   'P' },
#endif //shenzheng 2015-4-28 proxy administer
#if 1 //shenzheng 2015-6-8 zookeeper
#ifdef NC_ZOOKEEPER
	{ "zk-start",     		no_argument,  		NULL,   'S' },
	{ "zk-keep",     		no_argument,  		NULL,   'K' },
	{ "zk-server",     		required_argument,  NULL,   'z' },
	{ "zk-path",     		required_argument,  NULL,   'Z' },
#endif
#endif //shenzheng 2015-6-8 zookeeper
    { NULL,             	0,                  NULL,    0  }
};

#if 1 //shenzheng 2015-1-26 log rotating && proxy administer && zookeeper
#if 1 //shenzheng 2015-6-18 zookeeper
#ifdef NC_ZOOKEEPER
static char short_options[] = "hVtdDv:o:c:s:i:a:p:m:RM:C:A:P:z:Z:SK";
#else
static char short_options[] = "hVtdDv:o:c:s:i:a:p:m:RM:C:A:P:";
#endif
#else //shenzheng 2015-6-18 zookeeper
static char short_options[] = "hVtdDv:o:c:s:i:a:p:m:RM:C:A:P:z:Z:SK";
#endif
#else
static char short_options[] = "hVtdDv:o:c:s:i:a:p:m:";
#endif //shenzheng 2015-4-28 log rotating && proxy administer && zookeeper

static rstatus_t
nc_daemonize(int dump_core)
{
    rstatus_t status;
    pid_t pid, sid;
    int fd;

    pid = fork();
    switch (pid) {
    case -1:
        log_error("fork() failed: %s", strerror(errno));
        return NC_ERROR;

    case 0:
        break;

    default:
        /* parent terminates */
        _exit(0);
    }

    /* 1st child continues and becomes the session leader */

    sid = setsid();
    if (sid < 0) {
        log_error("setsid() failed: %s", strerror(errno));
        return NC_ERROR;
    }

    if (signal(SIGHUP, SIG_IGN) == SIG_ERR) {
        log_error("signal(SIGHUP, SIG_IGN) failed: %s", strerror(errno));
        return NC_ERROR;
    }

    pid = fork();
    switch (pid) {
    case -1:
        log_error("fork() failed: %s", strerror(errno));
        return NC_ERROR;

    case 0:
        break;

    default:
        /* 1st child terminates */
        _exit(0);
    }

    /* 2nd child continues */

    /* change working directory */
    if (dump_core == 0) {
        status = chdir("/");
        if (status < 0) {
            log_error("chdir(\"/\") failed: %s", strerror(errno));
            return NC_ERROR;
        }
    }

    /* clear file mode creation mask */
    umask(0);

    /* redirect stdin, stdout and stderr to "/dev/null" */

    fd = open("/dev/null", O_RDWR);
    if (fd < 0) {
        log_error("open(\"/dev/null\") failed: %s", strerror(errno));
        return NC_ERROR;
    }

    status = dup2(fd, STDIN_FILENO);
    if (status < 0) {
        log_error("dup2(%d, STDIN) failed: %s", fd, strerror(errno));
        close(fd);
        return NC_ERROR;
    }

    status = dup2(fd, STDOUT_FILENO);
    if (status < 0) {
        log_error("dup2(%d, STDOUT) failed: %s", fd, strerror(errno));
        close(fd);
        return NC_ERROR;
    }

    status = dup2(fd, STDERR_FILENO);
    if (status < 0) {
        log_error("dup2(%d, STDERR) failed: %s", fd, strerror(errno));
        close(fd);
        return NC_ERROR;
    }

    if (fd > STDERR_FILENO) {
        status = close(fd);
        if (status < 0) {
            log_error("close(%d) failed: %s", fd, strerror(errno));
            return NC_ERROR;
        }
    }

    return NC_OK;
}

static void
nc_print_run(struct instance *nci)
{
    int status;
    struct utsname name;

    status = uname(&name);
    if (status < 0) {
        loga("nutcracker-%s started on pid %d", NC_VERSION_STRING, nci->pid);
    } else {
        loga("nutcracker-%s built for %s %s %s started on pid %d",
             NC_VERSION_STRING, name.sysname, name.release, name.machine,
             nci->pid);
    }

    loga("run, rabbit run / dig that hole, forget the sun / "
         "and when at last the work is done / don't sit down / "
         "it's time to dig another one");
}

static void
nc_print_done(void)
{
    loga("done, rabbit done");
}

static void
nc_show_usage(void)
{
    log_stderr(
        "Usage: nutcracker [-?hVdDt] [-v verbosity level] [-o output file]" CRLF
        "                  [-c conf file] [-s stats port] [-a stats addr]" CRLF
        "                  [-i stats interval] [-p pid file] [-m mbuf size]" CRLF
        "");
    log_stderr(
        "Options:" CRLF
        "  -h, --help                : this help" CRLF
        "  -V, --version             : show version and exit" CRLF
        "  -t, --test-conf           : test configuration for syntax errors and exit" CRLF
        "  -d, --daemonize           : run as a daemon" CRLF
        "  -D, --describe-stats      : print stats description and exit");
    log_stderr(
        "  -v, --verbosity=N         : set logging level (default: %d, min: %d, max: %d)" CRLF
        "  -o, --output=S            : set logging file (default: %s)" CRLF
        "  -c, --conf-file=S         : set configuration file (default: %s)" CRLF
        "  -s, --stats-port=N        : set stats monitoring port (default: %d)" CRLF
        "  -a, --stats-addr=S        : set stats monitoring ip (default: %s)" CRLF
        "  -i, --stats-interval=N    : set stats aggregation interval in msec (default: %d msec)" CRLF
        "  -p, --pid-file=S          : set pid file (default: %s)" CRLF
        "  -m, --mbuf-size=N         : set size of mbuf chunk in bytes (default: %d bytes)" CRLF
        "",
        NC_LOG_DEFAULT, NC_LOG_MIN, NC_LOG_MAX,
        NC_LOG_PATH != NULL ? NC_LOG_PATH : "stderr",
        NC_CONF_PATH,
        NC_STATS_PORT, NC_STATS_ADDR, NC_STATS_INTERVAL,
        NC_PID_FILE != NULL ? NC_PID_FILE : "off",
        NC_MBUF_SIZE
		);

#if 1 //shenzheng 2015-1-26 log rotating
		log_stderr(
		"  -R, --log-rorate          : enable log rorate" CRLF
		"  -M, --log-file-max-size=N : set max size per log file if log-rorate open (default: %ld bytes, min: %d bytes, max: %ld bytes)" CRLF
        "  -C, --log-file-count=N    : set total count of log files to leave if log-rorate open (default: %d, min: %d, max: %d)"
		"",
		NC_LOG_FILE_MAX_SIZE_DEFAULT, NC_LOG_FILE_MAX_SIZE_MIN, NC_LOG_FILE_MAX_SIZE_MAX,
        NC_LOG_FILE_COUNT_DEFAULT, NC_LOG_FILE_COUNT_MIN, NC_LOG_FILE_COUNT_MAX
		);
#endif //shenzheng 2015-1-26 log rotating

#if 1 //shenzheng 2015-4-28 proxy administer
		log_stderr(
		"  -A, --proxy-adm-addr=S    : set proxy administer monitoring ip (default: %s)" CRLF
        "  -P, --proxy-adm-port=N    : set proxy administer monitoring port (default: %d, means disabled)"
        "",
        NC_PROXY_ADM_ADDR, NC_PROXY_ADM_PORT
		);
#endif //shenzheng 2015-4-28 proxy administer

#if 1 //shenzheng 2015-6-8 zookeeper
#ifdef NC_ZOOKEEPER
		log_stderr(
		"  -S, --zk-start            : set configuration from zookeeper" CRLF
        "  -K, --zk-keep             : set configuration keep with zookeeper" CRLF
        "  -z, --zk-server=S         : set zookeeper servers address (default: %s)" CRLF
        "  -Z, --zk-path=S           : set zookeeper configuration path (default: %s)"
        "",
        NC_ZOOKEEPER_ADDR, NC_ZOOKEEPER_PATH
		);
#endif
#endif //shenzheng 2015-6-8 zookeeper

#if 1 //shenzheng 2015-6-8 common
		log_stderr("");
#endif //shenzheng 2015-6-8 common


}

static rstatus_t
nc_create_pidfile(struct instance *nci)
{
    char pid[NC_UINTMAX_MAXLEN];
    int fd, pid_len;
    ssize_t n;

    fd = open(nci->pid_filename, O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (fd < 0) {
        log_error("opening pid file '%s' failed: %s", nci->pid_filename,
                  strerror(errno));
        return NC_ERROR;
    }
    nci->pidfile = 1;

    pid_len = nc_snprintf(pid, NC_UINTMAX_MAXLEN, "%d", nci->pid);

    n = nc_write(fd, pid, pid_len);
    if (n < 0) {
        log_error("write to pid file '%s' failed: %s", nci->pid_filename,
                  strerror(errno));
        return NC_ERROR;
    }

    close(fd);

    return NC_OK;
}

static void
nc_remove_pidfile(struct instance *nci)
{
    int status;

    status = unlink(nci->pid_filename);
    if (status < 0) {
        log_error("unlink of pid file '%s' failed, ignored: %s",
                  nci->pid_filename, strerror(errno));
    }
}

static void
nc_set_default_options(struct instance *nci)
{
    int status;

    nci->ctx = NULL;

    nci->log_level = NC_LOG_DEFAULT;
    nci->log_filename = NC_LOG_PATH;

    nci->conf_filename = NC_CONF_PATH;

    nci->stats_port = NC_STATS_PORT;
    nci->stats_addr = NC_STATS_ADDR;
    nci->stats_interval = NC_STATS_INTERVAL;

    status = nc_gethostname(nci->hostname, NC_MAXHOSTNAMELEN);
    if (status < 0) {
        log_warn("gethostname failed, ignored: %s", strerror(errno));
        nc_snprintf(nci->hostname, NC_MAXHOSTNAMELEN, "unknown");
    }
    nci->hostname[NC_MAXHOSTNAMELEN - 1] = '\0';

    nci->mbuf_chunk_size = NC_MBUF_SIZE;

    nci->pid = (pid_t)-1;
    nci->pid_filename = NULL;
    nci->pidfile = 0;
#if 1 //shenzheng 2015-1-26 log rotating
	LOG_RORATE = LOG_ROTATE_DEFAULT;
	LOG_FIEL_MAX_SIZE_FOR_ROTATING = NC_LOG_FILE_MAX_SIZE_DEFAULT;
	LOG_FILE_COUNT_TO_STAY = NC_LOG_FILE_COUNT_DEFAULT;
#endif //shenzheng 2015-1-26 log rotating

#if 1 //shenzheng 2015-4-28 proxy administer
	nci->proxy_adm_addr = NC_PROXY_ADM_ADDR;
	nci->proxy_adm_port = NC_PROXY_ADM_PORT;
#endif //shenzheng 2015-4-28 proxy administer

#if 1 //shenzheng 2015-6-9 zookeeper
#ifdef NC_ZOOKEEPER
	nci->zk_start = NC_ZOOKEEPER_START_DEFAULT;
	nci->zk_keep = NC_ZOOKEEPER_KEEP_DEFAULT;
	nci->zk_servers = NC_ZOOKEEPER_ADDR;
	nci->zk_path = NC_ZOOKEEPER_PATH;
#endif
#endif //shenzheng 2015-6-9 zookeeper

}

static rstatus_t
nc_get_options(int argc, char **argv, struct instance *nci)
{
    int c, value;

    opterr = 0;

    for (;;) {
        c = getopt_long(argc, argv, short_options, long_options, NULL);
        if (c == -1) {
            /* no more options */
            break;
        }

        switch (c) {
        case 'h':
            show_version = 1;
            show_help = 1;
            break;

        case 'V':
            show_version = 1;
            break;

        case 't':
            test_conf = 1;
            break;

        case 'd':
            daemonize = 1;
            break;

        case 'D':
            describe_stats = 1;
            show_version = 1;
            break;

        case 'v':
            value = nc_atoi(optarg, strlen(optarg));
            if (value < 0) {
                log_stderr("nutcracker: option -v requires a number");
                return NC_ERROR;
            }
            nci->log_level = value;
            break;

        case 'o':
            nci->log_filename = optarg;
            break;

        case 'c':
            nci->conf_filename = optarg;
            break;

        case 's':
            value = nc_atoi(optarg, strlen(optarg));
            if (value < 0) {
                log_stderr("nutcracker: option -s requires a number");
                return NC_ERROR;
            }
            if (!nc_valid_port(value)) {
                log_stderr("nutcracker: option -s value %d is not a valid "
                           "port", value);
                return NC_ERROR;
            }

            nci->stats_port = (uint16_t)value;
            break;

        case 'i':
            value = nc_atoi(optarg, strlen(optarg));
            if (value < 0) {
                log_stderr("nutcracker: option -i requires a number");
                return NC_ERROR;
            }

            nci->stats_interval = value;
            break;

        case 'a':
            nci->stats_addr = optarg;
            break;

        case 'p':
            nci->pid_filename = optarg;
            break;

        case 'm':
            value = nc_atoi(optarg, strlen(optarg));
            if (value <= 0) {
                log_stderr("nutcracker: option -m requires a non-zero number");
                return NC_ERROR;
            }

            if (value < NC_MBUF_MIN_SIZE || value > NC_MBUF_MAX_SIZE) {
                log_stderr("nutcracker: mbuf chunk size must be between %zu and"
                           " %zu bytes", NC_MBUF_MIN_SIZE, NC_MBUF_MAX_SIZE);
                return NC_ERROR;
            }

            nci->mbuf_chunk_size = (size_t)value;
            break;

#if 1 //shenzheng 2015-1-26 log rotating
		case 'R':
			LOG_RORATE = 1;
			break;
		case 'M':
		{
			char *result;
			result = set_log_file_max_size(optarg);
			if(result != NULL)
			{
				log_stderr("nutcracker: %s", result);
				nc_free(result);
				return NC_ERROR;
			}
			break;
		}
		case 'C':
		{
			char *result;
			result = set_log_file_count(optarg);
			if(result != NULL)
			{
				log_stderr("nutcracker: %s", result);
				nc_free(result);
				return NC_ERROR;
			}
			break;
		}
#endif //shenzheng 2015-1-26 log rotating

#if 1 //shenzheng 2015-4-28 proxy administer
		case 'A':
            nci->proxy_adm_addr = optarg;
            break;

		case 'P':
            value = nc_atoi(optarg, strlen(optarg));
            if (value < 0) {
                log_stderr("nutcracker: option -S requires a number");
                return NC_ERROR;
            }
            if (!nc_valid_port(value)) {
                log_stderr("nutcracker: option -S value %d is not a valid "
                           "port", value);
                return NC_ERROR;
            }

            nci->proxy_adm_port = (uint16_t)value;
            break;
#endif //shenzheng 2015-4-28 proxy administer

#if 1 //shenzheng 2015-6-8 zookeeper
#ifdef NC_ZOOKEEPER
		case 'z':
			nci->zk_servers = optarg;
			break;

		case 'Z':
			nci->zk_path = optarg;
			break;

		case 'S':
			nci->zk_start = 1;
			break;

		case 'K':
			nci->zk_keep = 1;
			break;
#endif
#endif //shenzheng 2015-6-8 zookeeper

        case '?':
            switch (optopt) {
            case 'o':
            case 'c':
            case 'p':
                log_stderr("nutcracker: option -%c requires a file name",
                           optopt);
                break;

            case 'm':
            case 'v':
            case 's':
            case 'i':
                log_stderr("nutcracker: option -%c requires a number", optopt);
                break;

            case 'a':
                log_stderr("nutcracker: option -%c requires a string", optopt);
                break;

#if 1 //shenzheng 2015-1-27 log rotating
			case 'M':
				log_stderr("nutcracker: option -%c requires a file size", optopt);
				break;
			case 'C':
				log_stderr("nutcracker: option -%c requires a number", optopt);
				break;
#endif //shenzheng 2015-1-27 log rotating

#if 1 //shenzheng 2015-4-28 proxy administer
			case 'A':
				log_stderr("nutcracker: option -%c requires a string", optopt);
				break;
			case 'P':
				log_stderr("nutcracker: option -%c requires a number", optopt);
				break;
#endif //shenzheng 2015-4-28 proxy administer

#if 1 //shenzheng 2015-6-8 zookeeper
#ifdef NC_ZOOKEEPER	
			case 'z':
				log_stderr("nutcracker: option -%c requires a string", optopt);
				break;
	
			case 'Z':
				log_stderr("nutcracker: option -%c requires a string", optopt);
				break;
#endif
#endif //shenzheng 2015-6-8 zookeeper

            default:
                log_stderr("nutcracker: invalid option -- '%c'", optopt);
                break;
            }
            return NC_ERROR;

        default:
            log_stderr("nutcracker: invalid option -- '%c'", optopt);
            return NC_ERROR;

        }
    }

    return NC_OK;
}

/*
 * Returns true if configuration file has a valid syntax, otherwise
 * returns false
 */
static bool
nc_test_conf(struct instance *nci)
{
    struct conf *cf;

    cf = conf_create(nci->conf_filename);
    if (cf == NULL) {
        log_stderr("nutcracker: configuration file '%s' syntax is invalid",
                   nci->conf_filename);
        return false;
    }

    conf_destroy(cf);

    log_stderr("nutcracker: configuration file '%s' syntax is ok",
               nci->conf_filename);
    return true;
}

static rstatus_t
nc_pre_run(struct instance *nci)
{
    rstatus_t status;

    status = log_init(nci->log_level, nci->log_filename);
    if (status != NC_OK) {
        return status;
    }

    if (daemonize) {
        status = nc_daemonize(1);
        if (status != NC_OK) {
            return status;
        }
    }

    nci->pid = getpid();

    status = signal_init();
    if (status != NC_OK) {
        return status;
    }

    if (nci->pid_filename) {
        status = nc_create_pidfile(nci);
        if (status != NC_OK) {
            return status;
        }
    }

    nc_print_run(nci);

    return NC_OK;
}

static void
nc_post_run(struct instance *nci)
{
    if (nci->pidfile) {
        nc_remove_pidfile(nci);
    }

    signal_deinit();

    nc_print_done();

    log_deinit();
}

static void
nc_run(struct instance *nci)
{
    rstatus_t status;
    struct context *ctx;

    ctx = core_start(nci);
    if (ctx == NULL) {
        return;
    }

    /* run rabbit run */
    for (;;) {
        status = core_loop(ctx);
        if (status != NC_OK) {
            break;
        }
    }

    core_stop(ctx);
}

int
main(int argc, char **argv)
{
    rstatus_t status;
    struct instance nci;

    nc_set_default_options(&nci);

    status = nc_get_options(argc, argv, &nci);
    if (status != NC_OK) {
        nc_show_usage();
        exit(1);
    }

    if (show_version) {
        log_stderr("This is nutcracker-%s" CRLF, NC_VERSION_STRING);
        if (show_help) {
            nc_show_usage();
        }

        if (describe_stats) {
            stats_describe();
        }

        exit(0);
    }

    if (test_conf) {
        if (!nc_test_conf(&nci)) {
            exit(1);
        }
        exit(0);
    }

    status = nc_pre_run(&nci);
    if (status != NC_OK) {
        nc_post_run(&nci);
        exit(1);
    }

    nc_run(&nci);

    nc_post_run(&nci);

    exit(1);
}
